containers.go

```diff
@@ -378,11 +382,22 @@ func (r *containerStore) removeName(container *Container, name string) {
 }
 
 func (r *containerStore) SetNames(id string, names []string) error {
-	names = dedupeNames(names)
+	opts := SetNameOptions{
+		AppendOnly: false,
+	}
+	return r.SetNamesWithOptions(id, names, opts)
+}
+
+func (r *containerStore) SetNamesWithOptions(id string, names []string, opts SetNameOptions) error {
 	if container, ok := r.lookup(id); ok {
 		for _, name := range container.Names {
-			delete(r.byname, name)
+			if opts.AppendOnly {
+				names = append(names, name)
+			} else {
+				delete(r.byname, name)
+			}
 		}
+		names = dedupeNames(names)
 		for _, name := range names {
 			if otherContainer, ok := r.byname[name]; ok {
```

Comment:
This looks quite broken: Not deleting `container` from `r.byname` will set `otherContainer` to `container` AFAICS.

I think that’s ultimately hidden by the `container.Names` assignment below, but it’s just not a clean and consistent way to implement this.

---

Same in the other two implementations.

---

One way to do this would be not to change the basic structure, only to change what names are set. I.e. keep the original `delete(r.byname)` loop as is, _then_ compute the final `names` value, and then proceed with the `otherContainer/byname/…` updates using the existing logic.

That would have the advantage that `updatedNames := applyEdits(container.Names, inputNames, opts)` could be shared across the three implementations.

Another way is to indeed deeply embed the add/remove cases into the implementation, so that names that don’t change are neither removed nor added in `r.byname`. That’s _technically_ more efficient but IMHO not worth the complexity.

Reply:
… actually, with a clean `addNames` / `removeNames` (were we do need to handle no-op names but they are only no-op additions / no-op removals) the “deeply embed” approach also might be simple enough.

Reply:
Yeah this part is not needed. Let me alter that, lets discuss new in original thread.